French translation and GNU/Linux packaging

DevDef opened the pull request:
Here’s my French translation and GNU/Linux packaging additions.
Enjoy :-)

kaikramer commented:
Well done, I will do a review of your changes over the weekend, but it looks really good at a first glance. 

I am also thankful for the little fixes you did besides translation, even for "cleaning up the mess". :smiley:

DevDef commented:
You’re welcome. :-)

Actually, there’s a lot of files that deserve a quick clean up of the “mess” with a mix of Windows and UNIX end of lines (almost all properties files, except the French one, of course...), a mix of tabs and spaces as indent and even useless trailing spaces.
I forgot to add in my commits that I corrected the build.xml ant file to use your new version.properties file which contains some strings parsed by ant which is then using them as “substitution variables”.

Concerning my Git pull, I still worry about the French properties files encoding, because I firstly tried to make a Git diff (from Eclipse) and the result was transforming all my accented characters to the same character (�). I think it’s because the diff file is UTF-8 encoded, whereas the properties files have to be ISO-8859-1, aka latin1, with Unicode escaped sequences (for example, \u2019 is the real apostrophe). I hope that the Git pull will work better...

DevDef commented:
I’ve finally made an SVG icon for KSE.

DevDef commented:
Ooops, closed the pull request by accident...

DevDef commented:
librsvg rendering is different than Inskape’s: the magnifier handle is brown in Inscape, but is displayed yellow with librsvg, the library used by GNOME Shell and eog. Not that bad, but not what I wanted.

kaikramer commented:
Alright, very well done. I am impressed. Must have been a lot of work. Did you do it all alone or did someone (Luc?) help you?

![french1](https://cloud.githubusercontent.com/assets/7840192/13549182/bf93e9f2-e300-11e5-8a1b-7e2eb01d026e.png)

I will accept this pull request as it is, but you can take your time to improve the SVG further if you want to.

DevDef commented:
I’m glad to see my entire pull request accepted.

Yes, I’ve done the translation all by myself. By the way, it’s my first translation for a Java project. Considering the amount of work it took me, I’m pretty proud of myself ;-). It would have been good to have someone looking at my inevitable mistakes and typos, but I didn’t dare asking someone for such a huge job...
Unfortunately, the translation work is not really fully completed yet, it still remains the help. Do you know the way to integrate the translation of these files? I don’t promise I’ll do it this time, but if I’m not too lazy, perhaps I’ll try in the long run...

Concerning the spec file for the RPM build, I provided it to ease the job of other RPM packagers. I think it could be used to build KSE for Fedora, RHEL/CentOS, SUSE, Mandriva, etc., with very few modifications, mostly adjusting the requirements and build requirements.

I’ll ask for inclusion into the upcoming Mageia 6. I think, it’s in a sufficient good shape now, notably since you fixed issue #11. I’m not an official packager, but they know me on the dev mailing list, and I should be able to convince Mageia’s people. Anyway, I’ll let you know.

I’m also a member of LinuxFr.org, a French speaking web site about free software. I plan the write something about KSE: how good it is, how nice the developer is and how productive our collaboration was. ;-) I hope some official packagers will be interested in packaging KSE for their favorite GNU/Linux distributions.

kaikramer commented:
Sounds great. I would love to see KSE included in Linux distributions. So far I only know of a request for inclusion in Debian from 2013 (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=729079).

As for the help files: It is probably not worth translating them right now, because I am not very happy with the current state of the help files for several reasons:

- The framework that is used to display them (javahelp) is very old and not maintained anymore.
- While it is a lot of text and covers almost every GUI element, it only describes what is there instead of guiding the user through complex tasks or even telling him where to start.
- The useful information is hidden between lots of rather superfluous text.
- Some of the help pages are just copies of the KSE website (like introduction and specifications) or of the license texts and I don't like redundancy.

So I am considering to replace the help pages from the KSE distribution with an online help (maybe even a wiki) that is more task/problem oriented in the near future.

DevDef commented:
OK, I forget the help for now. To be continued...

In the purpose of writing an article about KSE, could you please give me some information about it, you and the other developer? My questions are:
- Who initiated the project and why?
- Who are you (age, location, job)? If you agree to unveil personal information.- Is it a fork of Portecle or a complete rewrite with the same spirit?
- How long the work was until the first usable release?
- Are you the only developer now? Do you need help? 
- What do you plan for future releases? I’d like to see GTK file chooser integration, we already talked about it and I know that you agree

kaikramer commented:
Most of these questions are answered in this series of blog entries:

[A History of KeyStore Explorer - Part One](https://waynedgrant.wordpress.com/2011/12/22/a-history-of-keystore-explorer-part-one/)
[A History of KeyStore Explorer - Part Two](https://waynedgrant.wordpress.com/2012/02/21/a-history-of-keystore-explorer-part-two/)
[A History of KeyStore Explorer - Part Three](https://waynedgrant.wordpress.com/2012/05/18/a-history-of-keystore-explorer-part-three/)

Portecle is a fork of KSE, not the other way round. :smiley: 

There have been small code contributions by users, but apart from that I am the only developer right now. I understand that the skill set for working on KSE is very special and I don't expect or need help with coding in general. That being said, patches are always welcome and there are other ways to help because coding takes only about half of the time I spend for working on KSE. As I outlined in my comment above, the documentation needs some work for example. 

Often, when I see users writing about what they have done with KSE, I wonder why they are using only a very small set of its features. Like for example "I have used this openssl command to create a certificate and then I have used KSE to convert it to a Java keystore" or "I have used that keytool command to create a keystore with a certificate and then I have checked the content with KSE". That's a bit of a waste because KSE has excellent capabilities for creating certificates. A more task-oriented documentation should help with that.

My plans for future releases can be divided in three areas:

- Features of keytool or jarsigner that are still missing from KSE. After the release of KSE 5.2 (introducing support for PKCS#11, MSCAPI and timestamping) there won't be much left in this area (mainly jar signature verification and maybe support for some proprietary HSM keystore types). 
- "CA features": KSE is already pretty good here, but has for example some limitations regarding distinguished names (https://github.com/kaikramer/keystore-explorer/issues/13). Pre-defined templates for different certificate types (root CA, sub CA, SSL client, SSL server etc.) would be nice as well. CRL generation is missing completely. Release 5.3 will focus on these features.
- Usability

GTK file chooser integration is already done. It just requires a runtime with JavaFX.

I would like to keep my personal information private.

DevDef commented:
Thank you for the links about KSE’s history. I’ll read this soon...
I completely respect your desire of keeping private your personal information. However, could you make a little exception about the country you live in (and eventually precise the state, if you live in a federal country)? I hope not to be too intrusive.

I haven’t understood that JavaFX was providing the GTK file chooser. The good news is that it has been packaged for the upcoming Mageia 6 as java-1.8.0-openjfx, but not on my Mageia 5 (too many dependencies for a cherry pick of the package alone). I should add  a « Recommends: java-1.8.0-openjfx » in the spec file to do the trick for Mageia 6. Unless there is a need to add something to the classpath...

kaikramer commented:
Germany

Installing the java-1.8.0-openjfx package should suffice, adding the FX library to the classpath was only necessary for Java 7.

DevDef commented:
That Java FX library seems to be great, but I still have to see the result...

So, we are neighbors! :-) I thought you were from the other side of the Atlantic.
By the way, have you ever planned to translate KSE in German?

kaikramer commented:
Yep, neighbors indeed. :-) 

I have considered translating KSE, but the technical terms like "Common Name" or "Authority Key Identifier" are usually taken over anyway (only Microsoft tried to translate them into German and the result is awkward and confusing) and the rest is not so hard to understand, especially for the IT guys who usually use KSE. 

> That Java FX library seems to be great, but I still have to see the result...

Does this mean you have tried and it's not working?

DevDef commented:
> Yep, neighbors indeed. :-)

So, we share the same time zone (CET), it’s now 2:00 am. Isn’t it a bit late to work? ;-)

> I have considered translating KSE, but the technical terms like "Common Name" or "Authority Key Identifier" are usually taken over anyway (only Microsoft tried to translate them into German and the result is awkward and confusing) and the rest is not so hard to understand, especially for the IT guys who usually use KSE.

Perhaps you can put the English words between parenthesis when it sounds confusing:
« Gemeinsamen Namen (Common Name) », or just « Gemeinsamen Namen (CN) ». At least in the tooltips...

> > That Java FX library seems to be great, but I still have to see the result...
>
> Does this mean you have tried and it's not working?

No, I just meant that I don’t have Java FX packaged on my Mageia 5. As I told you, I tried to cherry pick the packages of the upcoming Mageia 6, but there are too many dependencies and I don’t want to break my system at the moment... So, I have three options: try to build Java FX on my current system, install a Mageia 6 on a virtual machine, or just wait a couple of months until Mageia 6 is stable enough to upgrade my system...

kaikramer commented:
> So, we share the same time zone (CET), it’s now 2:00 am. Isn’t it a bit late to work? ;-)

It was before 1:00 am when I wrote that comment, which is not late at all. :wink: 

> Perhaps you can put the English words between parenthesis when it sounds confusing:
« Gemeinsamen Namen (Common Name) », or just « Gemeinsamen Namen (CN) ». At least in the tooltips...

"Gemeinsamer Name" makes no sense at all. :smile: More correct would be "Gebräuchlicher Name" or "Allgemeiner Name" and that sounds stupid as well. I would probably just write "Name (CN)". 

Anyways, if someone else wants to translate KSE to German, I am fine with that, but I have a lot of other things to work on that seem more important to me.

> No, I just meant that I don’t have Java FX packaged on my Mageia 5. As I told you, I tried to cherry pick the packages of the upcoming Mageia 6, but there are too many dependencies and I don’t want to break my system at the moment... So, I have three options: try to build Java FX on my current system, install a Mageia 6 on a virtual machine, or just wait a couple of months until Mageia 6 is stable enough to upgrade my system...

Ok, I see.

DevDef commented:
I’ve finally successfully set up a Mageia 6 on virtual machine. I have small fixes to send you, so please don’t release yet.

TIA,

Davy

kaikramer commented:
You can relax, it will take at least another four weeks before the release is ready. 

A possible problem that occured to me just a few days ago: KSE 5.2 has an automatic update check. Basically the manual check (Help -> Online Resources -> Check for Update) is executed right after the start every 2 weeks. Not sure if this makes sense if KSE is installed as an rpm. A build switch might be necessary to disable this. What do you think?

BTW, would it be possible to move the spec file, the patch file and the other stuff that is required for building the rpm into their own sub-directory (maybe called "rpm")?

DevDef commented:
> You can relax, it will take at least another four weeks before the release is ready.

OK, I guess that should be more than enough to finish the polishing.

> A possible problem that occurred to me just a few days ago: KSE 5.2 has an automatic update check. Basically the manual check (Help -> Online Resources -> Check for Update) is executed right after the start every 2 weeks. Not sure if this makes sense if KSE is installed as an RPM. A build switch might be necessary to disable this. What do you think?

I totally agree that as an RPM, no update should be processed a all. And it would indeed be nice to switch it off at build.

> BTW, would it be possible to move the spec file, the patch file and the other stuff that is required for building the rpm into their own sub-directory (maybe called "rpm")?

Of course, that was my first intention. Let me do it for you...
BTW, I’m trying to nuke the patch because, in Mageia 6, the miglayout RPM provides miglayout-core.jar and miglayout-swing.jar instead of a single miglayout.jar. So it fits to the build.xml. It only remains javahelp2 dependency that is provided as javahelp2.jar by all the RPM based distros, but as jhall.jar by the Debian based distros, so this has to remain an exception. But I’d like to handle this with a build switch hence the patch could be removed. I have to see how to make a build switch, perhaps you could give me a hint...

Concerning JavaFX, I’ve been able to see the GTK file chooser, but it seems that we have indeed a problem with translations that I really want to be fixed before the 5.3 release.

kaikramer commented:
Alright, I have opted for a runtime switch (instead of build time) for disabling the update check. If the system property "_kse.packaging_" is set to "_rpm_", then the automatic update check is disabled and the menu item for doing it manually is removed. So you just have to set `-Dkse.packaging=rpm` in the start script.

As for the javahelp jar: It should be possible to modify the build.xml in a way that it uses whatever currently is in the lib directory instead of enumerating every single jar. It gets a bit tricky because of the manifest file, but if it works, it is the best solution IMHO.

kaikramer commented:
Oh, and I have removed half a sentence in one of the resource bundles.

DevDef commented:
> Alright, I have opted for a runtime switch (instead of build time) for disabling the update check. If the system property "kse.packaging" is set to "rpm", then the automatic update check is disabled and the menu item for doing it manually is removed. So you just have to add -Dkse.packaging=rpm in the start script

I like the idea, but not the name, because it’s not only for RPM packaging, but every GNU/Linux packaging and even more. I thing it should rather be a more specific name like "kse.update" and values true/false, 0/1.

> As for the javahelp jar: It should be possible to modify the build.xml in a way that it uses whatever currently is in the lib directory instead of enumerating every single jar. It gets a bit tricky because of the manifest file, but if it works, it is the best solution IMHO.

I just hope that it’s not a too big work for you...

By the way, I’ve found a little translation problem with JavaFX file chooser. Look at the "All files" (this is translated when not using JavaFX...
![capture d ecran de 2016-03-27 23 15 40](https://cloud.githubusercontent.com/assets/12715256/14067967/e43afbc6-f474-11e5-988f-0d15d907e319.png)
) in the attached screenshot...

DevDef commented:
One more little thing: I proposed you to try to unlock the system CA keystore with the default password "changeit" before prompting the user to enter a password, because it is rarely modified and most people doesn’t even know what this default password is. WDYT?

kaikramer commented:
Alright, but it has to default to "update check enabled" when missing, so I've renamed it to "kse.update.disabled" (with `-Dkse.update.disabled=true` to be effective).

I am really reluctant to add any new features at this point, however small they might be. There should be only bugfixes happening right now. I have implemented the "cacerts has moved to a new location" enhancement and I like the suggestion to try "changeit" before prompting the user, but if I don't find the time next week for it, then it simply won't make it into 5.2. Also, while it is right that most people don't know the password for cacerts, it is such a common problem, that a short google search quickly leads to the information.

DevDef commented:
> Alright, but it has to default to "update check enabled" when missing, so I've renamed it to "kse.update.disabled" (with -Dkse.update.disabled=true to be effective).

Perfect

> I am really reluctant to add any new features at this point, however small they might be. There should be only bugfixes happening right now. I have implemented the "cacerts has moved to a new location" enhancement and I like the suggestion to try "changeit" before prompting the user, but if I don't find the time next week for it, then it simply won't make it into 5.2. Also, while it is right that most people don't know the password for cacerts, it is such a common problem, that a short google search quickly leads to the information.

I understand, but that would be so smart and easy to do that I’d love to see it included in 5.2. :crying_cat_face:

You forgot to tell me you fixed the "All Files" translation. Thanks!

Now, I have to fight with git again! :weary:

kaikramer commented:
Okay, I've implemented it, but only because you have done so much for the project.

Maybe this video helps a bit with your git merge problems (it did help me): 
https://www.youtube.com/watch?v=KfeqnernMmE

DevDef commented:
Thank you so much Kai-san! :-)

DevDef commented:
It seems that some little changes have to be made somewhere else to complete the work... :-/
For example, when I import a trusted certificate, KSE is trying to check whether the certificate chain is valid or not, and it relies on the root certificates stored in the CA keystore. The problem is that KSE is prompting the user for the CA keystore password, the new code trying to unlock the CA keystore with the default password is not called. That’s really a pity.

Could you fix that too, please? TIA.
